primitives/src/rescue/sponge.rs

```diff
@@ -0,0 +1,441 @@
+// Copyright (c) 2022 Espresso Systems (espressosys.com)
+// This file is part of the Jellyfish library.
+
+// You should have received a copy of the MIT License
+// along with the Jellyfish library. If not, see <https://mit-license.org/>.
+
+//! This file contains the APIs wrappers for ark-sponge
+
+use ark_ff::{BigInteger, PrimeField};
+use ark_sponge::{
+    Absorb, CryptographicSponge, FieldBasedCryptographicSponge, FieldElementSize, SpongeExt,
+};
+use ark_std::{string::ToString, vec, vec::Vec};
+use jf_utils::{field_switching, pad_with_zeros};
+use num_bigint::BigUint;
+
+use super::{errors::RescueError, Permutation, RescueParameter, RescueVector, RATE, STATE_SIZE};
+
+#[derive(Clone, Default)]
+/// A rescue hash function consists of a permutation function and
+/// an internal state.
+pub struct RescueSponge<F: RescueParameter, const CHUNK_SIZE: usize> {
+    pub(crate) state: RescueVector<F>,
+}
+
+/// CRHF
+pub type RescueCRH<F> = RescueSponge<F, RATE>;
+/// PRF
+pub type RescuePRF<F> = RescueSponge<F, STATE_SIZE>;
```

Comment:
Now I tend to think that we should use newtype or some wrapper struct here instead of type alias. 

`RescueCRH` should follow a normal CRH API (something [like this](https://github.com/arkworks-rs/crypto-primitives/blob/main/src/crh/mod.rs)) and internally it can use all the `RescueSponge`'s API. and lives in `primitives/src/rescue/crh.rs` probably.

same for `RescuePRF`.

Maybe we can do this in a separate PR.

Reply:
Newtype probably makes sense.
I think in this PR I could still refactor it a bit more to separate out CRH and PRF (former having `sponge_with_padding` & `sponge_no_padding` methods, latter with `full_state_keyed_sponge_with_padding` & `full_state_keyed_sponge_no_padding`.

And potentially in a separate follow-up we could make our API follow the crypto-primitives trait from arkworks you 
linked.

WDYT? Also @EspressoSystems/jellyfish